### ISSUE TITLE
Transactions on CockroachDB violated Causal Consistency

### ISSUE THREAD

**Issue by 20211202na** [cockroachdb/cockroach#76436]
### Anomaly Description

We found transactional causal consistency (TCC) anomalies on CockroachDB. According to https://www.cockroachlabs.com/docs/v21.2/transactions.html , CockroachDB guarantees Serializability that is stronger than TCC (see its definition below).

We set up our experiment on a local machine with a single server node. Here is the configuration information:

CockroachDB Version == 21.2.5

server_node = 1
client_num(session_num) = 2 with multi-threading
txn_per_session = 10
operation_per_txn = 10
key_number = 10
key_distribution = uniform

We are using a simple table schema, containing key-value pairs. Here is the initialized table:

|**KEY** | **VALUE** |
|  ----  | ----  |
| 0 | 0 |
| 1 | 0 |
| 2 | 0 |
| ... | 0 |
| 9 | 0 |

Note that, for each write on a key, the value (generated by our workload generator) is unique. 

One anomaly was found on five transactions from two client sessions, where r/w(A,B) denotes that an operation read or writes value B on key A, respectively:

|**session 0** | **session 1** |
|  ----  | ----  |
|  txn0<br>w(8, 1) <br>r(4, 48)  |  txn16<br>w(8, 51)<br>w(4, 48)  |
| txn1<br>r(8, 51)  |  |

txn16 and txn1 have "write-read" order on key8 , denoted by txn16 ->wr txn1; txn16 and txn0 have write-read order on key 4, i.e., txn16 -> txn0.  Note that the "write-read" order relates two transactions that contain write/read operations on the same key-value pair; the transaction with the write operation should happen before the transaction with the read operation. 

To satisfy TCC, txn0 must be ordered before txn16 because we already know txn16 ->wr txn1 and txn0 ->so txn1(session order; ops/txns are closed loop). Thus we have 'commit order' txn0 ->co txn16 (see below for the definition of commit order). However, txn16 can reach txn0 via txn16 ->wr txn60. Hence, there is a cycle between txn16 and txn0 that violates TCC. 

![未命名文件](https://user-images.githubusercontent.com/96566054/153623608-76f96f41-cd45-4b0f-8eea-d5e5f93797b6.jpg)

The full dataset w.r.t. this counterexample is attached [data.txt](https://github.com/cockroachdb/cockroach/files/8049759/data.txt)

### Causal Consistency Definition
The causal consistency we checked is defined in the SOSP'11 paper (https://dl.acm.org/doi/10.1145/2043556.2043593). The authors defined a data consistency model called causal+. Causal+ is a combination of causal consistency and convergent conflict handling. Causal+ is strictly weaker than serializability. Regarding convergent conflict handling, the POPL'17 paper (https://dl.acm.org/doi/10.1145/3093333.3009888) has a very nice explanation:
“There is a total order between non-causally dependent operations and each site can execute operations only in that order (when it sees them). Therefore, a site is not allowed to revise its ordering of non-causally dependent operations, and all sites execute in the same order the operations that are visible to them.” 

### How to test and reproduce issue

1. We use  https://github.com/20211202na/cockroach_data/blob/main/cockroach_data_generation.py  to deal with the client-side logic (e.g., generating txn workloads)

2. For each run, “histories” of txns (values read and/or written) for all clients are also collected and printed.

3. We then invoke the checker (the function run_oopsla_graph) in https://github.com/20211202na/cockroach_data/blob/main/run_verification.py to verify if the history violates TCC.

Note that, since we are doing random testing (e.g., workloads are generated probabilistically), it’s hard (or even impossible) to reproduce a specific anomaly. However, we observed 1 violating history per 150 histories with the setup posted. We believe that anomalies manifest with sufficiently large number of runs. Additionally, more concurrency is expected to give more anomalies, e.g., with more clients, less keys, more txns, more ops per txn, etc.

**Comment by 20211202na** [comment]
Moreover, we found that this example also violates Read Atomicity (RA), an even weaker isolation level. Basically, with RA, either all or none of a transaction's updates are visible to another transaction's reads. The following hierarchy of isolation models was taken from Bailis et al. TODS paper https://dl.acm.org/doi/10.1145/2909870

![image](https://user-images.githubusercontent.com/96566054/153630845-dc57405a-d8dc-4342-913f-e75e331730a2.png)

**Comment by ajwerner** [comment]
Thanks for the report. What you're seeing definitely should not be possible in cockroach. I took a brief glance at your operation generator and something stuck out to me:

https://github.com/20211202na/cockroach_data/blob/7ed7c6578a4d87a5cc8ce7f407c2875b55049191/cockroach_data_generation.py#L269-L301

If you look here, you add to result_ops *before* the client has issued the `COMMIT` operation (which happens after leaving the `with` block)

```python
                    result_ops.append(temp_tx_op)
                    count += 1
```

If an error occurs (and one does seem very likely in some of these scenarios), then the code will swallow that exception rather than retry the transaction. May I suggest that you rewrite your code to not update the history or count until *after* a successful commit?

```python
    while True:
        if count == trans_num:
            break
        temp_tx_op = []
        try:
            with client_conn:
                with client_conn.cursor() as cur:
                    for j in range(0, len(list_of_ops[count])):
                        #             print('now dealing with : '+list_of_ops[i][j])
                        op = str.split(list_of_ops[count][j], ',', 3)
                        # create a new variable
                        key = str(op[1])
                        if (op[0] == 'write'):
                            val = str(op[2])
                            # print('now write, key is '+key+', value is '+val+'.\n')
                            op_update(cur, key, val)
                            single_op = 'w(' + str(key) + ',' + str(val) + ',' + str(client_id) + ',' + str(
                                count) + ',' + str(
                                op_num) + ')'
                            # print(single_op)
                            temp_tx_op.append(single_op)
                            op_num = op_num + 1
                        elif (op[0] == 'read'):
                            # print('now read, key is '+key+'.\n')
                            value = op_query(cur, key)
                            single_op = 'r(' + str(key) + ',' + str(value) + ',' + str(client_id) + ',' + str(
                                count) + ',' + str(
                                op_num) + ')'
                            # print(single_op)
                            temp_tx_op.append(single_op)
                            op_num = op_num + 1
                        else:
                            print("unknown wrong type op.")
        except Exception as e:
            continue
        result_ops.append(temp_tx_op)
        count += 1
```

**Comment by ajwerner** [comment]
Given this doesn't seem like a cockroach bug, I'm inclined to close it. We can continue the conversation here, if there's more to say. Otherwise, if some new problem arises from your research, please do open a new issue.

This sort of research on testing methodologies to discover consistency violations is very important. Hope this issue does not discourage your continued work in any way!

**Comment by ajwerner** [comment]
What you're missing is that commit operations can fail. Cockroach (and indeed many databases) check isolation properties upon COMMIT. 

> When a connection exits the with block, if no exception has been raised by the block, the transaction is committed.

The language here may be confusing. It's saying that upon leaving the with block, the driver will attempt to commit the transaction. It does not guarantee that the commit will succeed. 

I encourage you to try the code suggestion I typed above and to modify it to log the exceptions, including when the exceptions follow the completion of all the statements. I'm fairly certain you'll see exceptions indicating serializability violations. 

In general, swallowing exceptions without checking their type is a bad practice. In this specific case, assuming that because no exception was raised when issuing the statements of a transaction means that not exception will be raising when attempting to commit the transaction is just wrong.

**Comment by ajwerner** [comment]
> The official document, as well as the illustrating examples, is vague, especially w.r.t. multi-with-statements within a connection. Here, when it says "the transaction IS committed", how would one know that it actually means "trying to commit"?? 

I agree that the `psycopg2` documentation here should be improved. The documentation in question lives here: https://github.com/psycopg/psycopg2/blob/master/doc/src/connection.rst

Consider filing an issue or making a change there?
